### ISSUE TITLE
OSD can't be compiled out of the code

### ISSUE THREAD

**Issue by peterbarker** [ArduPilot/ardupilot#19071]
## Bug report

**Issue details**

OSD can't be compiled out:

```
define OSD_ENABLED 0
```

```
../../libraries/AP_MSP/AP_MSP_Telem_DJI.cpp: In member function 'virtual MSP::MSPCommandResult AP_MSP_Telem_DJI::msp_process_out_esc_sensor_data(MSP::sbuf_t*)':
../../libraries/AP_MSP/AP_MSP_Telem_DJI.cpp:105:36: error: 'class AP_OSD' has no member named 'get_stats_info'
  105 |         highest_temperature = osd->get_stats_info().max_esc_temp;
      |                                    ^~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
```

_Please describe the problem_

**Version**
master

**Platform**
[ X ] All
[  ] AntennaTracker
[  ] Copter
[  ] Plane
[  ] Rover
[  ] Submarine

**Airframe type**
n/a

**Hardware type**
NucleoH743

**Logs**
above

**Comment by amilcarlucas** [comment]
For flight software certification, it makes sense to be able to remove all non used features.
So I vote that this should get fixed. :)

**Comment by shiv-tyagi** [comment]
@peterbarker I tried to fix this. After compiling out the code causing problems in compilation, I am now getting symbol not found error while linking the files when I set `OSD_ENABLED 0`. However, I do not get such error when I set `OSD_PARAM_ENABLED 0`. 
Shouldn't we treat `OSD_PARAM_ENABLED` a dependency of `OSD_ENABLED` and check if OSD is enabled before we check OSD_PARAM_ENABLED wherever needed?

**Comment by andyp1per** [comment]
Doing this is really hard. There are many combinations that need to be accounted for.

**Comment by peterbarker** [comment]
On Mon, 1 Nov 2021, Shiv Tyagi wrote:

> @peterbarker I tried to fix this. After compiling out the code causing problems in compilation, I am now getting symbol not found error while linking the files when I set OSD_ENABLED 0. However, I do not get such error when I set OSD_PARAM_ENABLED 0.
> Shouldn't we treat OSD_PARAM_ENABLED a dependency of OSD_ENABLED and check if OSD is enabled before we check OSD_PARAM_ENABLED wherever needed?

Probably - you can see how that works out.

This stuff is all kinds of twisty as we re-use code in some weird ways :-)

**Comment by shiv-tyagi** [comment]
@peterbarker I have a PR (#19146) in which I tried to fix this. It would be great if you could review it and let me know if I need to do anything else to get this fixed.

**Comment by peterbarker** [comment]
@shiv-tyagi fixed this with #19146
